pkg/gameserversets/gameserversets.go

```diff
@@ -15,22 +15,83 @@
 package gameserversets
 
 import (
-	stablev1alpha1 "agones.dev/agones/pkg/apis/stable/v1alpha1"
+	"sort"
+
+	"agones.dev/agones/pkg/apis/stable/v1alpha1"
 	listerv1alpha1 "agones.dev/agones/pkg/client/listers/stable/v1alpha1"
 	"github.com/pkg/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/labels"
 )
 
+// node is just a convenience data structure for
+// keeping relevant GameServer information about Nodes
+type node struct {
+	name  string
+	total int64
+	ready []*v1alpha1.GameServer
+}
+
+// filterGameServersOnLeastFullNodes returns a limited list of GameServers, ordered by the nodes
+// they are hosted on, with the least utilised Nodes being prioritised
+func filterGameServersOnLeastFullNodes(gsList []*v1alpha1.GameServer, limit int32) []*v1alpha1.GameServer {
+	if limit <= 0 {
+		return nil
+	}
+
+	nodeMap := map[string]*node{}
+	var nodeList []*node
+
+	// count up the number of allocated and ready game servers that exist
+	// also, since we're already looping through, track all the deletable GameServers
+	// per node, so we can use this as a shortlist to delete from
+	for _, gs := range gsList {
+		if gs.DeletionTimestamp.IsZero() &&
+			(gs.Status.State == v1alpha1.Allocated || gs.Status.State == v1alpha1.Ready) {
+			_, ok := nodeMap[gs.Status.NodeName]
+			if !ok {
+				node := &node{name: gs.Status.NodeName}
+				nodeMap[gs.Status.NodeName] = node
+				nodeList = append(nodeList, node)
```

Comment:
doing memory allocations during a loop is not a good idea.
memory allocations are generally expensive, with a non-constant cost.
it's better to reserve memory in advance even if it's more than you need.
in this case, I suppose that the total number of nodes in the cluster should be already known locally?